File: lilac/src/main/java/com/hartwig/hmftools/lilac/LociPosition.java
package com.hartwig.hmftools.lilac;

import static com.hartwig.hmftools.common.gene.TranscriptUtils.calcCodingBases;
import static com.hartwig.hmftools.common.region.BaseRegion.positionWithin;

import com.google.common.collect.Lists;
import com.hartwig.hmftools.common.gene.ExonData;
import com.hartwig.hmftools.common.gene.TranscriptData;
import com.hartwig.hmftools.common.genome.bed.ImmutableNamedBed;
import com.hartwig.hmftools.common.genome.bed.NamedBed;

import java.util.List;

public class LociPosition
{
    private final List<TranscriptData> mTranscripts;

    public LociPosition(final List<TranscriptData> transcripts)
    {
        mTranscripts = transcripts;
    }

    public final int calcNucelotideLocus(int position)
    {
        for(TranscriptData transData : mTranscripts)
        {
            if(!positionWithin(position, transData.CodingStart, transData.CodingEnd))
                continue;

            // locus is a zero-based index, so the first coding base has locus of 0
            return calcCodingBases(transData, position).CodingBases - 1;
        }
        return -1;
    }
}

File: lilac/src/main/kotlin/com/hartwig/hmftools/lilac/LociPosition.kt
package com.hartwig.hmftools.lilac

import com.hartwig.hmftools.common.genome.region.HmfTranscriptRegion
import com.hartwig.hmftools.common.genome.region.Strand

class LociPosition(private val transcripts: List<HmfTranscriptRegion>) {

    fun position(codingLoci: Int): List<Int> {
        val result = mutableListOf<Int>()
        for (transcript in transcripts) {
            if (transcript.strand() == Strand.FORWARD) {
                result.add(forwardPosition(codingLoci, transcript))
            } else {
                result.add(reversePosition(codingLoci, transcript))
            }
        }
        return result
    }

    fun reversePosition(codingLoci: Int, transcript: HmfTranscriptRegion): Int {
        require(transcript.strand() == Strand.REVERSE)

        val codingStart = transcript.codingStart();
        val codingEnd = transcript.codingEnd();

        var currentLoci = 0
        for (exon in transcript.exome().reversed()) {
            if (exon.end() >= codingStart && exon.start() <= codingEnd) {
                val exonStartPosition = Math.max(codingStart, exon.start())
                val exonEndPosition = Math.min(codingEnd, exon.end())
                val exonLength = (exonEndPosition - exonStartPosition + 1).toInt()

                val exonStartLoci = currentLoci
                val exonEndLoci = exonStartLoci + exonLength - 1

                if (codingLoci in exonStartLoci..exonEndLoci) {
                    return (exonEndPosition - codingLoci + exonStartLoci).toInt()
                }

                currentLoci = exonEndLoci + 1
            }
        }

        return -1
    }

    fun forwardPosition(codingLoci: Int, transcript: HmfTranscriptRegion): Int {
        require(transcript.strand() == Strand.FORWARD)

        val codingStart = transcript.codingStart();
        val codingEnd = transcript.codingEnd();


        var currentLoci = 0
        for (exon in transcript.exome()) {
            if (exon.end() >= codingStart && exon.start() <= codingEnd) {
                val exonStartPosition = Math.max(codingStart, exon.start())
                val exonEndPosition = Math.min(codingEnd, exon.end())
                val exonLength = (exonEndPosition - exonStartPosition + 1).toInt()

                val exonStartLoci = currentLoci
                val exonEndLoci = exonStartLoci + exonLength - 1

                if (codingLoci in exonStartLoci..exonEndLoci) {
                    return (exonStartPosition + codingLoci - exonStartLoci).toInt()
                }

                currentLoci = exonEndLoci + 1
            }
        }

        return -1
    }

}